opencti-platform/opencti-graphql/src/database/engine-loader.ts

```diff
@@ -0,0 +1,1108 @@
+    }
+  }
+  // Return the elastic search engine expected bool should terms
+  // Build the search for all exact match (between double quotes)
+  const shouldSearch = [];
+  const cleanExactSearch = R.uniq(exactSearch.map((e) => e.replace(/"|http?:/g, '')));
+  shouldSearch.push(
+    ...cleanExactSearch.map((ex) => [
+      {
+        multi_match: {
+          type: 'phrase',
+          query: ex,
+          lenient: true,
+          fields: BASE_SEARCH_ATTRIBUTES,
+        },
+      },
+      {
+        nested: {
+          path: 'connections',
+          query: {
+            bool: {
+              must: [
+                {
+                  multi_match: {
+                    type: 'phrase',
+                    query: ex,
+                    lenient: true,
+                    fields: BASE_SEARCH_CONNECTIONS,
+                  },
+                },
+              ],
+            },
+          },
+        },
+      },
+    ]).flat()
+  );
+  // Build the search for all other fields
+  const searchPhrase = R.uniq(querySearch).join(' ');
+  if (searchPhrase) {
+    shouldSearch.push(...[
+      {
+        query_string: {
+          query: searchPhrase,
+          analyze_wildcard: true,
+          fields: BASE_SEARCH_ATTRIBUTES,
+        },
+      },
+      {
+        multi_match: {
+          type: 'phrase',
+          query: searchPhrase,
+          lenient: true,
+          fields: BASE_SEARCH_ATTRIBUTES,
+        },
+      },
+      {
+        nested: {
+          path: 'connections',
+          query: {
+            bool: {
+              must: [
+                {
+                  query_string: {
+                    query: searchPhrase,
+                    analyze_wildcard: true,
+                    fields: BASE_SEARCH_CONNECTIONS,
+                  },
+                },
+              ],
+            },
+          },
+        },
+      },
+    ]);
+  }
+  return shouldSearch;
+};
+
+export const buildUserMemberAccessFilter = (user: AuthUser, includeAuthorities = false) => {
+  const capabilities = user.capabilities.map((c) => c.name);
+  if (includeAuthorities && capabilities.includes(BYPASS)) {
+    return [];
+  }
+  const userAccessIds = computeUserMemberAccessIds(user);
+  // if access_users exists, it should have the user access ids
+  const authorizedFilters: any = [
+    { bool: { must_not: { exists: { field: 'authorized_members' } } } },
+    { terms: { 'authorized_members.id.keyword': [MEMBER_ACCESS_ALL, ...userAccessIds] } },
+  ];
+  if (includeAuthorities) {
+    const roleIds = user.roles.map((r) => r.id);
+    const owners = [...userAccessIds, ...capabilities, ...roleIds];
+    authorizedFilters.push({ terms: { 'authorized_authorities.keyword': owners } });
+  }
+  return [{ bool: { should: authorizedFilters } }];
+};
+
+export const elRawSearch = (context: AuthContext, user: AuthUser, types: string[], query: any) => {
```

Comment:
The functions that query the database share all the same problem regarding typescript conversion: we need to specify dynamically the input and output. Here, we use any and hope we find the right result according to our input.

The functions on the elastic client side are all generic, and what we do is some sort of wrapping around them to expose our own API ( `elPaginate`, `elQueryBuilder` etc.). Thus our functions should also be generic, and the code that calls them must define the input and output types locally, depending on the context.

Here again, this is a huge change, this time across all the backend. Probably not worth it right now.

Fortunately we have the graphql layer that is strictly typed, so the backend won't send something it's not supposed to send to the frontend. We could detect some problems that way (problems that would not be possible with proper TS typing).